lib/protocol/protocol.js

```diff
@@ -31,8 +31,65 @@ const IMG_EL_URL_RE = new RegExp(
 
 class Protocol {}
 
-function getLogByProtocol (driver) {
-  return driver.isW3CProtocol() ? w3cLog : mjsonwpLog;
+function getLogByProtocol (protocol) {
+  switch (protocol) {
+    case BaseDriver.DRIVER_PROTOCOL.W3C:
+      return w3cLog;
+    case BaseDriver.DRIVER_PROTOCOL.MJSONWP:
+      return mjsonwpLog;
+    default:
+      return genericProtocolLog;
+  }
+}
+
+// This cache is useful when a session gets terminated
+// and removed from the sessions list in the umbrella driver,
+// but the client still tries to send a command to the same session id.
+// So we know how to properly wrap the error message for it
+const PROTOCOLS_MAPPING_CACHE = {};
+const MAX_PROTOCOLS_MAPPING_CACHE_SIZE = 100;
+
+function extractProtocol (driver, sessionId = null) {
+  const cacheProtocolValue = (value) => {
+    if (!sessionId) {
+      return value;
+    }
+    if (PROTOCOLS_MAPPING_CACHE[sessionId]) {
+      return PROTOCOLS_MAPPING_CACHE[sessionId];
+    }
+    if (!value) {
+      return null;
+    }
+
+    PROTOCOLS_MAPPING_CACHE[sessionId] = [new Date().getTime(), value];
```

Comment:
i usually just use `Date.now()` cause it's short

Reply:
this is now obsolete